File: ldd-kit/adapters/nodejs/telemetry/tracing.ts
/**
 * OpenTelemetry tracing helpers for Node.js/TypeScript.
 *
 * Initialises a `NodeSDK`, exposes a `span` async wrapper for automatic
 * function instrumentation, and provides trace-context propagation via W3C
 * TraceContext.
 *
 * @module telemetry/tracing
 */

import { NodeSDK } from '@opentelemetry/sdk-node';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { ConsoleSpanExporter } from '@opentelemetry/sdk-trace-base';
import {
  trace,
  Tracer,
  Span,
  SpanKind,
  SpanStatusCode,
  context as otelContext,
  propagation,
  Context as OtelContext,
} from '@opentelemetry/api';

// ---------------------------------------------------------------------------
// Internal state
// ---------------------------------------------------------------------------

/** The globally initialised NodeSDK instance. */
let sdkInstance: NodeSDK | null = null;

/** Default tracer name used when no specific tracer is requested. */
let defaultTracerName = 'telemetry-default';

// ---------------------------------------------------------------------------
// SDK initialisation
// ---------------------------------------------------------------------------

/**
 * Initialise the global OpenTelemetry NodeSDK.
 *
 * If `otlpEndpoint` is provided, an OTLP HTTP exporter is configured. If not,
 * a `ConsoleSpanExporter` is used for development.
 *
 * **IMPORTANT**: You **must** call `shutdownTracing()` before process exit to
 * flush pending spans.
 *
 * @param serviceName - Value for the `service.name` resource attribute.
 * @param otlpEndpoint - OTLP HTTP endpoint URL (e.g.
 *   `http://otel-collector:4318/v1/traces`). If omitted, spans are printed to
 *   stdout.
 * @returns The configured NodeSDK instance.
 *
 * @example
 * ```ts
 * const sdk = initTracing('my-service', 'http://otel-collector:4318/v1/traces');
 * process.on('SIGTERM', () => sdk.shutdown());
 * ```
 */
export function initTracing(
  serviceName: string,
  otlpEndpoint?: string
): NodeSDK {
  const exporter = otlpEndpoint
    ? new OTLPTraceExporter({ url: otlpEndpoint })
    : new ConsoleSpanExporter();

  sdkInstance = new NodeSDK({
    serviceName,
    traceExporter: exporter,
  });

  sdkInstance.start();
  defaultTracerName = serviceName;

  return sdkInstance;
}

/**
 * Gracefully shut down the tracer provider, flushing any pending spans.
 *
 * @returns Promise that resolves when shutdown is complete.
 */
export async function shutdownTracing(): Promise<void> {
  if (sdkInstance) {
    await sdkInstance.shutdown();
    sdkInstance = null;
  }
}

// ---------------------------------------------------------------------------
// Tracer factory
// ---------------------------------------------------------------------------

/**
 * Return a named `Tracer` from the global provider.
 *
 * @param name - Tracer name — conventionally `__filename` or module name.
 * @returns A tracer instance.
 */
export function getTracer(name: string): Tracer {
  return trace.getTracer(name);
}

/**
 * Return the default tracer for the service.
 */
export function getDefaultTracer(): Tracer {
  return trace.getTracer(defaultTracerName);
}

// ---------------------------------------------------------------------------
// Span wrapper
// ---------------------------------------------------------------------------

/**
 * Execute an async function within a new OpenTelemetry span.
 *
 * The span is automatically ended when `fn` completes (success or failure).
 * If `fn` throws, the error is recorded on the span and the span status is set
 * to ERROR before re-throwing.
 *
 * @param name - Span operation name.
 * @param fn - Async (or sync) function to execute within the span. Receives
 *   the created Span for adding attributes and events.
 * @param kind - Optional span kind (default: `SpanKind.INTERNAL`).
 * @returns A Promise resolving to `fn`'s return value.
 *
 * @example
 * ```ts
 * const result = await span('process_image', async (span) => {
 *   span.setAttribute('image.size', data.length);
 *   return await processImage(data);
 * });
 * ```
 */
export async function span<T>(
  name: string,
  fn: (span: Span) => T | Promise<T>,
  kind: SpanKind = SpanKind.INTERNAL
): Promise<T> {
  const tracer = getDefaultTracer();

  return tracer.startActiveSpan(name, { kind }, async (span) => {
    try {
      const result = await fn(span);
      span.setStatus({ code: SpanStatusCode.OK });
      return result;
    } catch (err) {
      const errorMessage = err instanceof Error ? err.message : String(err);
      span.recordException(err instanceof Error ? err : new Error(errorMessage));
      span.setStatus({
        code: SpanStatusCode.ERROR,
        message: errorMessage,
      });
      throw err;
    } finally {
      span.end();
    }
  });
}

// ---------------------------------------------------------------------------
// Trace context propagation
// ---------------------------------------------------------------------------

/**
 * Inject the current span context into a headers object.
 *
 * Uses the W3C TraceContext propagator to serialise traceparent / tracestate.
 *
 * @param headers - Mutable headers object to inject into. Defaults to `{}`.
 * @returns The headers object with trace context injected.
 *
 * @example
 * ```ts
 * const headers = injectTraceContext({ 'content-type': 'application/json' });
 * await fetch('http://downstream/api', { headers });
 * ```
 */
export function injectTraceContext(
  headers: Record<string, string> = {}
): Record<string, string> {
  const setter = {
    set(carrier: Record<string, string>, key: string, value: string): void {
      carrier[key] = value;
    },
  };
  propagation.inject(otelContext.active(), headers, setter);
  return headers;
}

/**
 * Extract trace context from incoming headers and return an OpenTelemetry
 * Context carrying the extracted span.
 *
 * @param headers - Incoming headers containing traceparent.
 * @returns An OpenTelemetry Context with the extracted span, or the active
 *   context if extraction fails.
 */
export function extractTraceContext(headers: Record<string, string>): OtelContext {
  const getter = {
    keys(carrier: Record<string, string>): string[] {
      return Object.keys(carrier);
    },
    get(carrier: Record<string, string>, key: string): string | string[] | undefined {
      return carrier[key];
    },
  };
  return propagation.extract(otelContext.active(), headers, getter);
}

/**
 * Extract the trace ID from the current active span as a hex string.
 * Returns an empty string if no span is active.
 */
export function getTraceID(): string {
  const currentSpan = trace.getActiveSpan();
  if (!currentSpan || !currentSpan.spanContext().isValid) {
    return '';
  }
  return currentSpan.spanContext().traceId;
}

/**
 * Extract the span ID from the current active span as a hex string.
 * Returns an empty string if no span is active.
 */
export function getSpanID(): string {
  const currentSpan = trace.getActiveSpan();
  if (!currentSpan || !currentSpan.spanContext().isValid) {
    return '';
  }
  return currentSpan.spanContext().spanId;
}

File: ldd-kit/adapters/rust/telemetry/src/tracing.rs
//! OpenTelemetry tracing helpers for Rust.
//!
//! Initialises a `TracerProvider`, provides trace-context propagation via W3C
//! TraceContext, and integrates with the `tracing` crate via
//! `tracing-opentelemetry`.
//!
//! The `#[tracing::instrument]` attribute from the `tracing` crate provides
//! automatic span creation for functions. This module handles the global
//! tracer provider setup and context propagation.
//!
//! # Example
//!
//! ```no_run
//! use telemetry::init_tracing;
//! use tracing::{info, instrument};
//!
//! #[tokio::main]
//! async fn main() {
//!     let _provider = init_tracing("my-service", Some("http://otel-collector:4318"));
//!
//!     process_image(vec![0u8; 1024]).await;
//! }
//!
//! #[tracing::instrument(fields(image_size = data.len()))]
//! async fn process_image(data: Vec<u8>) {
//!     info!(result = "ok", "image processed");
//! }
//! ```

use opentelemetry::{
    global,
    propagation::TextMapPropagator,
    sdk::{
        propagation::TraceContextPropagator,
        resource::Resource,
        trace::{self, TracerProvider},
        runtime::Tokio,
    },
    trace::{TraceError, Tracer},
    KeyValue,
};
use opentelemetry_otlp::WithExportConfig;
use std::collections::HashMap;
use tracing_opentelemetry::OpenTelemetryLayer;
use tracing_subscriber::{layer::SubscriberExt, util::SubscriberInitExt, EnvFilter};

/// Initialise the global OpenTelemetry tracer provider.
///
/// If `otlp_endpoint` is provided, an OTLP HTTP exporter is configured. If
/// not, a stdout exporter is used for development.
///
/// # Arguments
///
/// * `service_name` — Value for the `service.name` resource attribute.
/// * `otlp_endpoint` — OTLP HTTP endpoint URL (e.g.
///   `"http://otel-collector:4318/v1/traces"`). If `None`, spans are printed
///   to stdout.
///
/// # Returns
///
/// The configured `TracerProvider`. This should be held for the lifetime of
/// the application and shut down on exit:
///
/// ```no_run
/// use telemetry::{init_tracing, shutdown_tracing};
///
/// #[tokio::main]
/// async fn main() {
///     let provider = init_tracing("my-service", None);
///     // ... application logic ...
///     shutdown_tracing();
/// }
/// ```
pub fn init_tracing(service_name: &str, otlp_endpoint: Option<&str>) -> TracerProvider {
    let resource = Resource::new(vec![KeyValue::new(
        opentelemetry_semantic_conventions::resource::SERVICE_NAME,
        service_name.to_string(),
    )]);

    let exporter: Box<dyn trace::SpanExporter + Send + Sync> = match otlp_endpoint {
        Some(endpoint) => {
            let otlp_exporter = opentelemetry_otlp::new_exporter()
                .http()
                .with_endpoint(endpoint.to_string());

            match otlp_exporter.build_span_exporter() {
                Ok(exporter) => Box::new(exporter),
                Err(e) => {
                    eprintln!(
                        "telemetry: OTLP exporter failed ({}), falling back to stdout",
                        e
                    );
                    Box::new(
                        trace::stdout::new_pipeline()
                            .with_trace_config(
                                trace::config().with_resource(resource.clone()),
                            )
                            .install_simple(),
                    )
                }
            }
        }
        None => Box::new(
            trace::stdout::new_pipeline()
                .with_trace_config(trace::config().with_resource(resource.clone()))
                .install_simple(),
        ),
    };

    let provider = TracerProvider::builder()
        .with_resource(resource)
        .with_batch_exporter(exporter, Tokio)
        .build();

    global::set_tracer_provider(provider.clone());

    // Set up the TraceContext propagator for W3C traceparent/tracestate.
    global::set_text_map_propagator(TraceContextPropagator::new());

    // Install the tracing-opentelemetry layer so that #[tracing::instrument]
    // spans are exported via OTel.
    let otel_layer = OpenTelemetryLayer::new(provider.tracer(service_name));

    // Only install the subscriber if one hasn't already been set.
    // This is a best-effort — in production, the application should set up
    // the full subscriber in main().
    let _ = tracing_subscriber::registry()
        .with(EnvFilter::try_from_default_env().unwrap_or_else(|_| EnvFilter::new("info")))
        .with(otel_layer)
        .try_init();

    provider
}

/// Get a named `Tracer` from the global provider.
///
/// # Arguments
///
/// * `name` — Tracer name — conventionally the module path.
pub fn get_tracer(name: &'static str) -> opentelemetry::global::BoxedTracer {
    global::tracer(name)
}

/// Shutdown the global tracer provider, flushing any pending spans.
///
/// Call this before your application exits.
pub fn shutdown_tracing() {
    global::shutdown_tracer_provider();
}

/// Inject the current span context from the active tracing span into the
/// provided headers map using the W3C TraceContext propagator.
///
/// # Arguments
///
/// * `headers` — Mutable `HashMap` to inject trace context into.
///
/// # Example
///
/// ```no_run
/// use telemetry::inject_trace_context;
/// use std::collections::HashMap;
///
/// let mut headers = HashMap::new();
/// headers.insert("content-type".to_string(), "application/json".to_string());
/// inject_trace_context(&mut headers);
/// // headers now contains traceparent and tracestate
/// ```
pub fn inject_trace_context(headers: &mut HashMap<String, String>) {
    let propagator = TraceContextPropagator::new();
    let mut injector = HeaderInjector { headers };
    propagator.inject_context(&tracing::Span::current().context(), &mut injector);
}

/// Extract trace context from incoming headers and return an OpenTelemetry
/// `Context` carrying the extracted span.
///
/// # Arguments
///
/// * `headers` — Incoming headers that may contain traceparent.
///
/// # Returns
///
/// An OpenTelemetry `Context` with the extracted span context, or an empty
/// context if extraction fails.
pub fn extract_trace_context(headers: &HashMap<String, String>) -> opentelemetry::Context {
    let propagator = TraceContextPropagator::new();
    let extractor = HeaderExtractor { headers };
    propagator.extract(&extractor)
}

/// Extract the trace ID from the current active span as a hex string.
/// Returns an empty string if no span is active.
pub fn trace_id() -> String {
    let span = tracing::Span::current();
    let ctx = span.context();
    let span_context = ctx.span().span_context();
    if span_context.is_valid() {
        span_context.trace_id().to_string()
    } else {
        String::new()
    }
}

/// Extract the span ID from the current active span as a hex string.
/// Returns an empty string if no span is active.
pub fn span_id() -> String {
    let span = tracing::Span::current();
    let ctx = span.context();
    let span_context = ctx.span().span_context();
    if span_context.is_valid() {
        span_context.span_id().to_string()
    } else {
        String::new()
    }
}

// ---------------------------------------------------------------------------
// Internal: HeaderInjector / HeaderExtractor for W3C propagation
// ---------------------------------------------------------------------------

/// Implements `Injector` for HashMap-based headers.
struct HeaderInjector<'a> {
    headers: &'a mut HashMap<String, String>,
}

impl<'a> opentelemetry::propagation::Injector for HeaderInjector<'a> {
    fn set(&mut self, key: &str, value: String) {
        self.headers.insert(key.to_lowercase(), value);
    }
}

/// Implements `Extractor` for HashMap-based headers.
struct HeaderExtractor<'a> {
    headers: &'a HashMap<String, String>,
}

impl<'a> opentelemetry::propagation::Extractor for HeaderExtractor<'a> {
    fn get(&self, key: &str) -> Option<&str> {
        self.headers
            .get(&key.to_lowercase())
            .map(|v| v.as_str())
    }

    fn keys(&self) -> Vec<&str> {
        self.headers.keys().map(|k| k.as_str()).collect()
    }
}
